tests_and_analysis/test/euphonic_test/test_spectrum1d.py

```diff
@@ -0,0 +1,223 @@
+        spec1d = Spectrum1D(*args, **kwargs)
+        return spec1d, expected_spec1d
+
+    @pytest.fixture(params=[
+        'quartz_666_dos.json',
+        'xsq_spectrum1d.json',
+        'xsq_bin_edges_spectrum1d.json'])
+    def create_from_json(self, request):
+        json_file = request.param
+        expected_spec1d = get_expected_spectrum1d(json_file)
+        spec1d = Spectrum1D.from_json_file(
+            get_json_file(json_file))
+        return spec1d, expected_spec1d
+
+    @pytest.fixture(params=[
+        'quartz_666_dos.json',
+        'xsq_spectrum1d.json',
+        'xsq_bin_edges_spectrum1d.json'])
+    def create_from_dict(self, request):
+        json_file = request.param
+        expected_spec1d = get_expected_spectrum1d(json_file)
+        spec1d = Spectrum1D.from_dict(
+            expected_spec1d.to_dict())
+        return spec1d, expected_spec1d
+
+    @pytest.mark.parametrize(('spec1d_creator'), [
+        pytest.lazy_fixture('create_from_constructor'),
+        pytest.lazy_fixture('create_from_json'),
+        pytest.lazy_fixture('create_from_dict')])
+    def test_correct_object_creation(self, spec1d_creator):
+        spec1d, expected_spec1d = spec1d_creator
+        check_spectrum1d(spec1d, expected_spec1d)
+
+    @pytest.fixture(params=[
+        ('x_data',
+         get_expected_spectrum1d('xsq_spectrum1d.json').x_data.magnitude,
+         TypeError),
+        ('x_data',
+         get_expected_spectrum1d('xsq_spectrum1d.json').x_data[:-1],
+         ValueError),
+        ('y_data',
+         get_expected_spectrum1d('xsq_spectrum1d.json').y_data.magnitude,
+         TypeError),
+        ('y_data',
+         get_expected_spectrum1d('xsq_spectrum1d.json').y_data[:-2],
+         ValueError),
+        ('x_tick_labels',
+         get_expected_spectrum1d('xsq_spectrum1d.json').x_tick_labels[0],
+         TypeError)])
+    def inject_faulty_elements(self, request):
+        faulty_arg, faulty_value, expected_exception = request.param
+        expected_spec1d = get_expected_spectrum1d('xsq_spectrum1d.json')
+        # Inject the faulty value and get a tuple of constructor arguments
+        args, kwargs = expected_spec1d.to_constructor_args(
+            **{faulty_arg: faulty_value})
+        return args, kwargs, expected_exception
+
+    def test_faulty_object_creation(self, inject_faulty_elements):
+        faulty_args, faulty_kwargs, expected_exception = inject_faulty_elements
+        with pytest.raises(expected_exception):
+            Spectrum1D(*faulty_args, **faulty_kwargs)
+
+
+@pytest.mark.unit
+class TestSpectrum1DSerialisation:
+
+    @pytest.fixture(params=[
+        get_spectrum1d('quartz_666_dos.json'),
+        get_spectrum1d('xsq_spectrum1d.json'),
+        get_spectrum1d('xsq_bin_edges_spectrum1d.json')])
+    def serialise_to_json_file(self, request, tmpdir):
+        spec1d = request.param
+        # Serialise
+        output_file = str(tmpdir.join('tmp.test'))
```

Comment:
The example in the Pytest docs uses it this way, so it's _probably_  fine? https://docs.pytest.org/en/stable/tmpdir.html

Reply:
The first example has `tmpdir` as an argument to _the test itself_ whereas here it is being passed to another fixture that is used by the test. I don't think that should make a difference, but it would be nice to ask a Pytest guru if we had one.

The docs go on to show how `tmpdir_factory` would be used to create a directory that _is_ re-used at the session (i.e. the behaviour I want to avoid.)

As the docs describe `tmpdir` as "the tmpdir fixture which will provide a temporary directory unique to the test invocation" it seems reasonable to assume we're ok as-is. But there will be a way of expressing the scope more explicitly in cases where this is delicate.

Happy to accept this as-is.